### PR TITLE
Make sure to establish connection and empty the event queue before cancelling publisher 

### LINKED ISSUE #7522: Scheduler is flaky, "Too many realizations failed"
**Describe the bug**
The test `tests/integration_tests/analysis/test_es_update.py::test_update_multiple_param` is very flaky, atleast using scheduler. 
```
FAILED tests/integration_tests/analysis/test_es_update.py::test_update_multiple_param[using_scheduler] - ert.cli.main.ErtCliError: Too many realizations have failed! Number of successful realizations: 9, expected minimal number of successful realizations: 10
```
(https://github.com/equinor/ert/actions/runs/8420486419/job/23055374718?pr=7521)

**Expected behaviour**
Not flaky.

**Environment**
 - OS: Mostly macos-latest
 - ERT/Komodo release: main
 - Python version: Mostly (if not only) 3.12
 - Remote/HPC execution involved: Github Actions runner

**Additional context**

### PR DESCRIPTION
**Pull request by xjules**
**Issue**
Resolves #7522 

- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Added appropriate release note label
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure tests pass locally (after every commit!)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [x] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: src/ert/scheduler/scheduler.py [hunk 1]
```diff
@@ -263,6 +263,8 @@ async def execute(
         try:
             await self._monitor_and_handle_tasks(scheduling_tasks)
         finally:
+            while not self._events.empty():
```
Comment (berland) [inline src/ert/scheduler/scheduler.py:266]: Maybe it is `_monitor_and_handle_tasks` that should not return until this is empty. 
And inside there one might consider if this should be a test inside `is_active()`, that scheduler should consider itself active while there are events pending.

Comment (xjules) [reply src/ert/scheduler/scheduler.py:266]: Let's give it a try then

Comment (berland) [reply src/ert/scheduler/scheduler.py:266]: simulation_context is using `is_active()` so maybe we should put the guard inside there to avoid the same problem popping up there.

Comment (berland) [reply src/ert/scheduler/scheduler.py:266]: I vote for changing `is_active()` to return True whenever the event queue is nonempty.

Comment (berland) [reply src/ert/scheduler/scheduler.py]: Getting there, I see we need to mock `websocket.client.connect` in order to test this.

Comment (berland) [reply src/ert/scheduler/scheduler.py]: And possibly we should only require the queue to become empty if `self.__ee_uri` is not None.

Comment (xjules) [reply src/ert/scheduler/scheduler.py]: I needed to change it back to a version where task handles cancellation itself. Two main reasons:
1) The way task is doing cleanup should be an implementation detail and as such we don't want to create class variables (eg. publisher running) and maintain them.
2) Simulation context handling becomes far more complex just to check publisher running and queue not empty - I mean the tests mostly.
Thus the best option would be to do the cleanup once we cancel the task as this would become an implementation detail of the publisher task.

File: src/ert/scheduler/scheduler.py [hunk 2]
```diff
@@ -228,9 +231,10 @@ async def _monitor_and_handle_tasks(
                         continue
                     if task_exception := task.exception():
                         raise task_exception
-                # make sure to empty report all events before exit
-                while not self._events.empty():
-                    await asyncio.sleep(0)
+                # make sure to report all events before exit
+                if self._publisher_task_active:
```
Comment (jonathan-eq) [inline src/ert/scheduler/scheduler.py]: Should we again set `self._publisher_task_active = False` after this if block is executed?

Comment (jonathan-eq) [reply src/ert/scheduler/scheduler.py]: I was looking at a specific commit rather than the entire change...

Comment (xjules) [reply src/ert/scheduler/scheduler.py]: Now it's gone - I've reverted it back as described in the comment https://github.com/equinor/ert/pull/7562#discussion_r1553047562

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -16,7 +16,7 @@ repos:
     - id: ruff-format
 
 -   repo: https://github.com/pre-commit/mirrors-clang-format
-    rev: v18.1.1
+    rev: v18.1.3
```
Comment (berland) [inline .pre-commit-config.yaml:19]: would be nice to have in a separate commit in case of rollbacks.

File: src/ert/async_utils.py [hunk 4]
```diff
@@ -40,6 +41,16 @@ def _done_callback(task: asyncio.Task[_T_co]) -> None:
         if (exc := task.exception()) is None:
             return
 
-        logger.error(f"Exception occurred during {task.get_name()}", exc_info=exc)
+        # logger.error(f"Exception occurred during {task.get_name()}", exc_info=exc)
```
Comment (berland) [inline src/ert/async_utils.py:44]: rm this line

File: .pre-commit-config.yaml (hunk 3 above)
Comment (xjules) [reply .pre-commit-config.yaml:19]: Sorry, left over. Will rm.

File: tests/integration_tests/analysis/test_es_update.py [hunk 5]
```diff
@@ -164,7 +164,6 @@ def sample_prior(nx, ny):
         )
 
 
-@pytest.mark.skip(reason="Very flaky with scheduler")
```
Comment (berland) [inline tests/integration_tests/analysis/test_es_update.py:167]: thank you for not forgetting this! 😄

File: src/ert/scheduler/scheduler.py [hunk 6]
```diff
@@ -215,14 +225,18 @@ async def _monitor_and_handle_tasks(
                     )
                     logger.error(
                         (
-                            f"Exception in scheduler task: {task_exception}\n"
+                            f"Exception in scheduler task {task.get_name()}: {task_exception}\n"
```
Comment (berland) [inline src/ert/scheduler/scheduler.py:228]: This should perhaps be a separate commit, as it is unrelated to fixing the premature closing of publisher.

Comment (xjules) [reply src/ert/scheduler/scheduler.py:228]: Now it's in two commits. 1) improving exception messages 2) the actual fix

**Review by berland — APPROVED** [review]
splendid
